File: src/main/java/me/miki/shindo/utils/BrowserUtils.java
package me.miki.shindo.utils;

import net.minecraft.util.Util;
import org.lwjgl.Sys;

import java.awt.Desktop;
import java.io.IOException;
import java.net.URI;

/**
 * Utility to open system browsers in a cross-platform way.
 */
public final class BrowserUtils {

    private BrowserUtils() {
    }

    public static boolean openUrl(String url) {
        if (url == null || url.trim().isEmpty()) {
            return false;
        }

        URI uri;
        try {
            uri = URI.create(url.trim());
        } catch (IllegalArgumentException ex) {
            return false;
        }

        if (tryDesktopBrowse(uri)) {
            return true;
        }

        if (tryNativeOpen(uri.toString())) {
            return true;
        }

        return tryLwjglFallback(uri.toString());
    }

    private static boolean tryDesktopBrowse(URI uri) {
        if (!Desktop.isDesktopSupported()) {
            return false;
        }

        try {
            Desktop desktop = Desktop.getDesktop();
            if (!desktop.isSupported(Desktop.Action.BROWSE)) {
                return false;
            }
            desktop.browse(uri);
            return true;
        } catch (IOException | SecurityException ignored) {
            return false;
        }
    }

    private static boolean tryNativeOpen(String url) {
        try {
            Util.EnumOS os = Util.getOSType();
            if (os == Util.EnumOS.OSX) {
                Runtime.getRuntime().exec(new String[]{"/usr/bin/open", url});
                return true;
            }

            if (os == Util.EnumOS.WINDOWS) {
                Runtime.getRuntime().exec(new String[]{"cmd.exe", "/C", "start", "\"\"", url});
                return true;
            }
        } catch (IOException ignored) {
        }

        return false;
    }

    private static boolean tryLwjglFallback(String url) {
        try {
            Sys.openURL(url);
            return true;
        } catch (Exception ignored) {
            return false;
        }
    }
}

File: src/main/kotlin/me/miki/shindo/utils/BrowserUtils.kt
package me.miki.shindo.utils

import net.minecraft.util.Util
import org.lwjgl.Sys
import java.awt.Desktop
import java.io.IOException
import java.net.URI
object BrowserUtils {

    @JvmStatic
    fun openUrl(url: String?): Boolean {
        if (url.isNullOrBlank()) {
            return false
        }

        val uri = try {
            URI.create(url.trim())
        } catch (_: IllegalArgumentException) {
            return false
        }

        if (tryDesktopBrowse(uri)) return true
        if (tryNativeOpen(uri.toString())) return true
        return tryLwjglFallback(uri.toString())
    }

    private fun tryDesktopBrowse(uri: URI): Boolean {
        if (!Desktop.isDesktopSupported()) return false

        return try {
            val desktop = Desktop.getDesktop()
            if (!desktop.isSupported(Desktop.Action.BROWSE)) return false
            desktop.browse(uri)
            true
        } catch (_: IOException) {
            false
        } catch (_: SecurityException) {
            false
        }
    }

    private fun tryNativeOpen(url: String): Boolean {
        return try {
            when (Util.getOSType()) {
                Util.EnumOS.OSX -> {
                    Runtime.getRuntime().exec(arrayOf("/usr/bin/open", url))
                    true
                }

                Util.EnumOS.WINDOWS -> {
                    Runtime.getRuntime().exec(arrayOf("cmd.exe", "/C", "start", "\"\"", url))
                    true
                }

                else -> false
            }
        } catch (_: IOException) {
            false
        }
    }

    private fun tryLwjglFallback(url: String): Boolean {
        return try {
            Sys.openURL(url)
            true
        } catch (_: Exception) {
            false
        }
    }
}
